src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,352 @@
+#if DEBUG
+                DebugObject.IsCurrentCompilation = true;
+#endif
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+
+                return GetAccessibleTopLevelTypesWorker(project.Id, compilation.Assembly.GlobalNamespace, checksum, excludedNamespaces, isInternalsVisible: true, _projectItemsCache, cancellationToken);
+            }
+
+            public async Task<ImmutableArray<TypeImportCompletionItem>> GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+                Solution solution,
+                Compilation compilation,
+                CompilationReference compilationReference,
+                ImmutableHashSet<string> excludedNamespaces,
+                CancellationToken cancellationToken)
+            {
+#if DEBUG
+                DebugObject.IsCurrentCompilation = false;
+#endif
+
+                if (!(compilation.GetAssemblyOrModuleSymbol(compilationReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return ImmutableArray<TypeImportCompletionItem>.Empty;
+                }
+
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var assemblyProject = solution.GetProject(assemblySymbol, cancellationToken);
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(assemblyProject, cancellationToken).ConfigureAwait(false);
+
+                return GetAccessibleTopLevelTypesWorker(assemblyProject.Id, assemblySymbol.GlobalNamespace, checksum, excludedNamespaces, isInternalsVisible, _projectItemsCache, cancellationToken);
+            }
+
+            public ImmutableArray<TypeImportCompletionItem> GetAccessibleTopLevelTypesFromPEReference(
+                Solution solution,
+                Compilation compilation,
+                PortableExecutableReference peReference,
+                ImmutableHashSet<string> excludedNamespaces,
+                CancellationToken cancellationToken)
+            {
+                if (!(compilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return ImmutableArray<TypeImportCompletionItem>.Empty;
+                }
+
+                var key = GetReferenceKey(peReference);
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var rootNamespaceSymbol = assemblySymbol.GlobalNamespace;
+
+                if (key == null)
+                {
+                    // Can't cache items for reference with null key, so just create them and return. 
+                    return GetCompletionItemsForTopLevelTypeDeclarations(rootNamespaceSymbol, ns => !excludedNamespaces.Contains(ns), isInternalsVisible);
```

Comment:
when do you get a null key?

Reply:
Not sure. But I copied the logic from SymbolTreeInfo

http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs,116